torch_geometric/utils/augmentation.py

```diff
@@ -0,0 +1,190 @@
+    edge indices.
+
+    Args:
+        edge_index (LongTensor): The edge indices.
+        p (float, optional): Ratio of added edges to the existing edges.
+            (default: :obj:`0.2`)
+        force_undirected (bool, optional): If set to :obj:`True`,
+            added edges will be undirected.
+            (default: :obj:`False`)
+        num_nodes (int, Tuple[int], optional): The overall number of nodes,
+            *i.e.* :obj:`max_val + 1`, or the number of source and
+            destination nodes, *i.e.* :obj:`(max_src_val + 1, max_dst_val + 1)`
+            of :attr:`edge_index`. (default: :obj:`None`)
+        training (bool, optional): If set to :obj:`False`, this operation is a
+            no-op. (default: :obj:`True`)
+
+    :rtype: (:class:`LongTensor`, :class:`LongTensor`)
+
+    Examples:
+
+        >>> # Standard case
+        >>> edge_index = torch.tensor([[0, 1, 1, 2, 2, 3],
+        ...                            [1, 0, 2, 1, 3, 2]])
+        >>> edge_index, added_edges = add_random_edge(edge_index, p=0.5)
+        >>> edge_index
+        tensor([[0, 1, 1, 2, 2, 3, 2, 1, 3],
+                [1, 0, 2, 1, 3, 2, 0, 2, 1]])
+        >>> added_edges
+        tensor([[2, 1, 3],
+                [0, 2, 1]])
+
+        >>> # The returned graph is kept undirected
+        >>> edge_index, added_edges = add_random_edge(edge_index, p=0.5,
+        ...                                           force_undirected=True)
+        >>> edge_index
+        tensor([[0, 1, 1, 2, 2, 3, 2, 1, 3, 0, 2, 1],
+                [1, 0, 2, 1, 3, 2, 0, 2, 1, 2, 1, 3]])
+        >>> added_edges
+        tensor([[2, 1, 3, 0, 2, 1],
+                [0, 2, 1, 2, 1, 3]])
+
+        >>> # For bipartite graphs
+        >>> edge_index = torch.tensor([[0, 1, 2, 3, 4, 5],
+        ...                            [2, 3, 1, 4, 2, 1]])
+        >>> edge_index, added_edges = add_random_edge(edge_index, p=0.5,
+        ...                                           num_nodes=(6, 5))
+        >>> edge_index
+        tensor([[0, 1, 2, 3, 4, 5, 3, 4, 1],
+                [2, 3, 1, 4, 2, 1, 1, 3, 2]])
+        >>> added_edges
+        tensor([[3, 4, 1],
+                [1, 3, 2]])
+    """
+    if p < 0. or p > 1.:
+        raise ValueError(f'Ratio of added edges has to be between 0 and 1 '
+                         f'(got {p}')
+
+    device = edge_index.device
+    if not training or p == 0.0:
+        edge_index_to_add = torch.tensor([], device=device).view(2, 0)
+        return edge_index, edge_index_to_add
+
+    num_nodes = (num_nodes,
+                 num_nodes) if not isinstance(num_nodes, tuple) else num_nodes
+    num_src_nodes = maybe_num_nodes(edge_index, num_nodes[0])
+    num_dst_nodes = maybe_num_nodes(edge_index, num_nodes[1])
+
+    num_edges_to_add = round(edge_index.size(1) * p)
+    row = torch.randint(0, num_src_nodes, size=(num_edges_to_add, ))
+    col = torch.randint(0, num_dst_nodes, size=(num_edges_to_add, ))
+
+    if force_undirected:
```

Comment:
I think `force_undirected` makes only sense for homogeneous graphs/in case `num_nodes` is an integer. In that case, we also probably want to only sample half of the edges.